### PR TITLE
Engagement validation for risk_acceptance API in POST PATCH and PUT

### PR DESCRIPTION
**Pull request by FelixHernandez**
## Engagement validation for risk_acceptance API in POST PATCH and PUT
[sc-4081]

**Description**

This feature validates the `Risk_Acceptance` creation. Only is allowed accept the `Finding` objects in the related `Engagement`

### CODE REVIEW

**Review by mtesauro — APPROVED** [review]
Approved

File: dojo/api_v2/serializers.py [hunk 1]
```diff
@@ -1531,13 +1532,27 @@ def get_engagement(self, obj):
         )
 
     def validate(self, data):
+        findings = data.get('accepted_findings', [])
+        findings_ids = list(map(lambda x: x.id, findings))
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:1536]:
```suggestion
        findings_ids = [x.id for x findings]
```
List comprehension would be easier to read here and also has less datatype conversions

File: dojo/api_v2/serializers.py [hunk 2]
```diff
@@ -1531,13 +1532,27 @@ def get_engagement(self, obj):
...
+        finding_objects = Finding.objects.filter(id__in=findings_ids)
+        authed_findings = get_authorized_findings(Permissions.Finding_View).filter(id__in=findings_ids)
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:1538]:
```suggestion
        authed_findings = get_authorized_findings(Permissions.Finding_Edit).filter(id__in=findings_ids)
```
Setting a risk exception should likely require at least edit access to a finding. Also removing extra var

File: dojo/api_v2/serializers.py [hunk 3]
```diff
@@ -1531,13 +1532,27 @@ def get_engagement(self, obj):
...
+                "You are not permitted to add one or more selected findings to this risk acceptance"
+            )
         if self.context["request"].method == "POST":
-            findings = data['accepted_findings']
-            for finding in findings:
-                if not user_has_permission(self.context["request"].user, finding, Permissions.Finding_View):
-                    raise PermissionDenied(
-                        "You are not permitted to add one or more selected findings to this risk acceptance"
-                    )
+            engagements = finding_objects.values_list('test__engagement__id', flat=True).distinct().count()
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:1544]:
```suggestion
            engagements = authed_findings.values_list('test__engagement__id', flat=True).distinct().count()
```

File: dojo/api_v2/serializers.py [hunk 4]
```diff
@@ -1531,13 +1532,27 @@ def get_engagement(self, obj):
...
+            if engagements > 1:
+                raise PermissionDenied(
+                    "You are not permitted to add findings to a distinct engagement"
+                )
+        elif self.context['request'].method in ['PATCH', 'PUT']:
+            engagement = Engagement.objects.filter(risk_acceptance=self.instance.id).first()
+            findings = finding_objects.exclude(test__engagement__id=engagement.id)
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:1551]:
```suggestion
            findings = authed_findings.exclude(test__engagement__id=engagement.id)
```

**Review by Maffooch — CHANGES_REQUESTED** [review]

File: dojo/api_v2/serializers.py (hunk 4 above)
Comment (FelixHernandez) [reply dojo/api_v2/serializers.py:1551]: `authed_findings` are the Findings the user have access to, `finding_objects` are the Finding the user is trying to accept with the `risk_acceptance`

**Review by grendel513 — APPROVED** [review]

**Review by devGregA — APPROVED** [review]

File: dojo/api_v2/serializers.py (hunk 3 above)
Comment (Maffooch) [reply dojo/api_v2/serializers.py:1544]: That's fine. They should be the same query at that point anyways :)

**Review by Maffooch — APPROVED** [review]

**Pull request merged**